src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,273 @@
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* function */
+      const char *get_name ();
+      void add_cur_stat (uint64_t size);
+      void sub_cur_stat (uint64_t size);
+      const uint64_t get_cur_stat();
+
+      ~mmm_subcomponent ()
+      {
+	delete[] m_subcomp_name;
+      }
+
+    private:
+      /* private variable */
+      char *m_subcomp_name;
+      std::atomic<uint64_t> m_cur_stat;
+  };
+
+  class mmm_component
+  {
+    public:
+      mmm_component (const char *name)
+      {
+	m_comp_name = new char[strlen (name) + 1];
+	strcpy (m_comp_name, name);
+      }
+
+      /* function */
+      const char *get_name ();
+      void add_init_stat (int size);
+      void sub_init_stat (int size);
+      void add_cur_stat (int size);
+      void sub_cur_stat (int size);
+      void add_peak_stat (int size);
+      void sub_peak_stat (int size);
+      void add_expand_count ();
+      const MMM_MEM_STAT &get_stat();
+      const std::vector<mmm_subcomponent *> &get_subcomp_vec ();
+      int add_subcomponent (const char *name);
+
+      ~mmm_component ()
+      {
+	delete[] m_comp_name;
+      }
+    private:
+      /* private variable */
+      char *m_comp_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_subcomponent *> m_subcomponent;
+  };
+
+  class mmm_module
+  {
+    public:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    public:
+      mmm_module (const char *name, const MMM_COMP_INFO *info);
+      mmm_module () {} /* for dummy */
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+      const char *get_name ();
+      void add_init_stat (int size);
+      void sub_init_stat (int size);
+      void add_cur_stat (int size);
+      void sub_cur_stat (int size);
+      void add_peak_stat (int size);
+      void sub_peak_stat (int size);
+      void add_expand_count ();
+      const MMM_MEM_STAT &get_stat ();
+      int add_component (const char *name);
+      const std::vector<mmm_component *> &get_comp_vec ();
+      const std::vector<int> &get_comp_idx_map ();
+      void add_comp_idx_map (int compound);
+
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      virtual ~mmm_module ()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+    private:
+      /* private variable */
+      char *m_module_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_component *> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  /* you have to add your modules in Memory Monitoring Manager(MMM) class */
+  class memory_monitoring_manager
+  {
+    public:
+      memory_monitoring_manager (const char *name)
```

Comment:
Seperates the declaration and implementation of the member function(include ctor, dtor)? It would help the code reader to easily understand the structures of the class because the class declaration is shorter, the overall structure is easier to figure out.

```
AS-IS
class A
{
         A()
         {
                 // 20 lines
         }
         func1()
         {
                // 100 lines
         }
    private:
         int a;
}

TO-BE
class A
{
         A()
         func1()
    private:
         int a;
}

// implement here
A::A()
{
           // 20 lines
}

A::func1()
{
           // 100 lines
}

Reply:
Done.

Reply:
But "static" functions just can be defined at class body. I change it to "static inline" to "inline".